File: javascript/find_shortest_string.js
function findShortestString(arr) {
  // type your code here

  let arrayOfLengths = arr.map(string => string.length)

  let shortest = Math.min(...arrayOfLengths)

  let shortestIndex = arrayOfLengths.indexOf(shortest)

  return arr[shortestIndex]
}

if (require.main === module) {
  // add your own tests in here
  console.log("Expecting: 'a'");
  console.log("=>", findShortestString(['aaa', 'a', 'bb', 'ccc']));

  console.log("Expecting: 'this'");
  console.log("=>", findShortestString(['hopefully', 'this', 'works', 'properly']));

  console.log("Expecting: 'hi'");
  console.log("=>", findShortestString(['cat', 'hi', 'dog', 'an']));

  console.log("Expecting: 'my'");
  console.log("=>", findShortestString(['my', 'fingers', 'are', 'crossed']));

  console.log("Expecting: 'lily'");
  console.log("=>", findShortestString(['flower', 'juniper', 'lily', 'dadelion']));

  // BENCHMARK HERE
}

module.exports = findShortestString;

// Please add your pseudocode to this file
  // map through array of strings, returning length of string
  // find minimum element in array of lengths and get index
  // return element in string array at that index

// And a written explanation of your solution
  // we create a new array that stores the lengths of all the strings, in order
  // we find the smallest number in that array of lengths and grab its index
  // then return the string at that index in the original array

File: ruby/find_shortest_string.rb
def find_shortest_string(arr)
  # type your code in here

  array_of_lengths = arr.map do |string|
    string.length
  end

  shortest = array_of_lengths.min

  shortest_index = array_of_lengths.index(shortest)

  arr[shortest_index]

end

if __FILE__ == $PROGRAM_NAME
  puts "Expecting: 'a'"
  puts "=>", find_shortest_string(['aaa', 'a', 'bb', 'ccc'])

  puts "Expecting: 'this'"
  puts "=>", find_shortest_string(['hopefully', 'this', 'works', 'properly'])

  puts "Expecting: 'hi'"
  puts "=>", find_shortest_string(['cat', 'hi', 'dog', 'an'])

  puts "Expecting: 'my'"
  puts "=>", find_shortest_string(['my', 'fingers', 'are', 'crossed'])

  puts "Expecting: 'lily'"
  puts "=>", find_shortest_string(['flower', 'juniper', 'lily', 'dadelion'])

  # Don't forget to add your own!

  # BENCHMARK HERE
end

# Please add your pseudocode to this file
# And a written explanation of your solution
